components/butterfly/src/rumor.rs

```diff
@@ -118,16 +117,29 @@ impl<'a, T: Rumor> From<&'a T> for RumorKey {
     fn from(rumor: &'a T) -> RumorKey { RumorKey::new(rumor.kind(), rumor.id(), rumor.key()) }
 }
 
+type ServiceGroupName = String;
+type MemberId = String;
+
 /// Storage for Rumors. It takes a rumor and stores it according to the member that produced it,
 /// and the service group it is related to.
 ///
 /// Generic over the type of rumor it stores.
 #[derive(Debug, Clone)]
 pub struct RumorStore<T: Rumor> {
-    pub list:       Arc<Lock<HashMap<String, HashMap<String, T>>>>,
+    list:           Arc<Lock<HashMap<ServiceGroupName, HashMap<MemberId, T>>>>,
```

Comment:
I don't think this is accurate.  I can see by checking the final diff that this particular line of code didn't stay this way, but technically the `RumorStore` models things as `HashMap<Rumor::key(), HashMap<Rumor::id(), Rumor>>`.  For the `Service` rumor, e.g. this formulation is correct because `Rumor::key()` returns the service group of the service and `Rumor::id()` returns the member ID.

For `Election` rumors, however, `Rumor::id()` returns the hardcoded string `election`, `ServiceConfig` returns the hardcoded string `service_config`.  For `Departure` rumors, `Rumor::key()` returns the hardcoded string `departure`, not the service group, and `Rumor::id()` returns the member ID.  There's quite a lot of variability.

As I said, I realize this code changed from what it is here, on this particular commit, to the end state, so I'll re-evaluate as I go.

Reply:
I think you're right about the naming. What would be preferable is *not* having every `RumorStore` be a two-level map that takes two keys (with inconsistent naming), and instead define it like:
```rust
pub struct RumorStore<S> {
    pub list:       Arc<RwLock<HashMap<String, S>>>,
    update_counter: Arc<AtomicUsize>,
}
```
Then, according to the needs of the particular rumor, `S` can be the rumor, or else another `HashMap`. I've hacked around with this a bit, and it's definitely possible, but I'm not sure if it's worth the effort involved.

Assuming we *don't* go that route, would you suggest naming these aliases differently, or getting rid of them altogether?

Reply:
I like your formulation with the single level map, but I agree that the effort involved there would be non-trivial.

As far as the naming goes, my gut reaction is to name it something like:

`HashMap<RumorKey, HashMap<RumorId, T>>`

only because those names map naturally to what's actually inserted into the `HashMap`.